trillian/integration/chains.go

```diff
@@ -0,0 +1,245 @@
+	reparsed, err := x509.ParseCertificate(precertData)
+	if err != nil {
+		return nil, fmt.Errorf("failed to re-parse created precertificate: %v", err)
+	}
+	return x509.BuildPrecertTBS(reparsed.RawTBSCertificate, preIssuer)
+}
+
+// makePreIssuerPrecertChain builds a precert chain where the pre-cert is signed by a new
+// pre-issuer intermediate.
+func makePreIssuerPrecertChain(chain []ct.ASN1Cert, issuer *x509.Certificate, signer crypto.Signer) ([]ct.ASN1Cert, []byte, error) {
+	prechain := make([]ct.ASN1Cert, len(chain)+1)
+	copy(prechain[2:], chain[1:])
+
+	// Create a new private key and intermediate CA cert to go with it.
+	preSigner, err := keys.NewFromSpec(&keyspb.Specification{
+		Params: &keyspb.Specification_EcdsaParams{
+			EcdsaParams: &keyspb.Specification_ECDSA{
+				Curve: keyspb.Specification_ECDSA_P256,
+			},
+		},
+	})
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create pre-issuer private key: %v", err)
+	}
+
+	preIssuerTemplate := *issuer
+	preIssuerTemplate.RawSubject = nil
+	preIssuerTemplate.Subject.CommonName += "PrecertIssuer"
+	preIssuerTemplate.PublicKeyAlgorithm = x509.ECDSA
+	preIssuerTemplate.PublicKey = preSigner.Public()
+	preIssuerTemplate.ExtKeyUsage = append(preIssuerTemplate.ExtKeyUsage, x509.ExtKeyUsageCertificateTransparency)
+
+	// Set a new subject-key-id for the intermediate (to ensure it's different from the true
+	// issuer's subject-key-id).
+	randData := make([]byte, 128)
+	if _, err := rand.Read(randData); err != nil {
+		return nil, nil, fmt.Errorf("failed to read random data: %v", err)
+	}
+	preIssuerTemplate.SubjectKeyId = randData
+	prechain[1].Data, err = x509.CreateCertificate(rand.Reader, &preIssuerTemplate, issuer, preIssuerTemplate.PublicKey, signer)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create pre-issuer certificate: %v", err)
+	}
+
+	// Parse the pre-issuer back to a fully-populated x509.Certificate.
+	preIssuer, err := x509.ParseCertificate(prechain[1].Data)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to re-parse generated pre-issuer: %v", err)
+	}
+
+	cert, err := x509.ParseCertificate(chain[0].Data)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to parse certificate to build precert from: %v", err)
+	}
+
+	prechain[0].Data, err = buildNewPrecertData(cert, preIssuer, preSigner)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create certificate: %v", err)
+	}
+
+	if err := verifyChain(prechain); err != nil {
+		return nil, nil, fmt.Errorf("failed to verify just-created prechain: %v", err)
+	}
+
+	// The leaf data has the poison removed and the issuance information changed.
+	tbs, err := buildLeafTBS(prechain[0].Data, preIssuer)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to build leaf TBSCertificate: %v", err)
+	}
+	return prechain, tbs, nil
+}
+
+// verifyChain checks that a chain of certificates validates locally.
+func verifyChain(rawChain []ct.ASN1Cert) error {
```

Comment:
Surprised there isn't existing code for this.

Reply:
Yeah, there probably should be a verifier in the client to help reduce the chances of people submitting invalid chains.